File: 2022/11.14/정호조/다리를지나는트럭.java
import java.util.*;

public class 다리를지나는트럭 {

	public static void main(String[] args) {

		int bridge_length = 100;
		int weight = 100;
		int[] truck_weights = {10, 10, 10, 10, 10, 10, 10, 10, 10, 10};
		Solution s = new Solution();
		s.solution(bridge_length, weight, truck_weights);

	}

	public static class Solution {

		public int solution(int bridge_length, int weight, int[] truck_weights) {
			
			int answer = Integer.MAX_VALUE;
			ArrayList<Integer> truck = new ArrayList<>();
			ArrayList<Car> curTruck = new ArrayList<>();
			
			for(int i=0; i<truck_weights.length; i++) {
				truck.add(truck_weights[i]);
			}
			
			int time = 0;
			int curWeight = 0;

			while (true) {
				if (truck.size() == 0 && curTruck.size() == 0) {
					break;
				}
				
				time++;
				
				if (curTruck.size() != 0 && time - curTruck.get(0).enterTime == bridge_length) {
					curWeight -= curTruck.get(0).weight;
					curTruck.remove(0);
				}
				
				if (truck.size() != 0 && curTruck.size() < bridge_length && curWeight < weight) {
					if (curWeight + truck.get(0) <= weight) {
						curTruck.add(new Car(truck.get(0), time));
						curWeight += truck.get(0);
						truck.remove(0);
					}
				}
			}

			answer = Math.min(answer, time);

			System.out.println(answer);
			return answer;
		}

		public static class Car {
			int weight, enterTime;

			public Car(int weight, int enterTime) {
				this.weight = weight;
				this.enterTime = enterTime;
			}
		}
	}	
}

File: 09.12/조수연/다리를 지나는 트럭.kt
package com.ssafy.algorithm


import java.util.*
import kotlin.collections.ArrayList

fun main() {
    val sl = Solution()
    val bridge_length = 100
    val weight = 100
    val truck_weights = intArrayOf(10,10,10,10,10,10,10,10,10,10)
    println(sl.solution(bridge_length, weight, truck_weights))

}

class Solution {
    lateinit var cross: BooleanArray //현재 트럭이 다리에 있는지 체크 배열
    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        var answer = 0
        val wait: Queue<Int> = LinkedList()
        val bridge: Queue<Int> = LinkedList()

        for (i in truck_weights){ //대기 트럭
            wait.offer(i)
        }

        var possible = 0
        var time = 0 //시간

        cross = BooleanArray(truck_weights.size)
        val trucks = IntArray(truck_weights.size)
        var index = 0
        while (true) {
            time++
            if (wait.isEmpty() && bridge.isEmpty()) break //대기 트럭이랑 다리에 트럭 큐가 비어있으면 종료

            if (wait.isNotEmpty()){ //대기트럭이 있을 때
                if (possible + wait.peek() <= weight) { //현재 다리위에 있는 트럭 + 트럭의 무게가 다리가 버틸 수 있는 무게라면 트럭추가
                    val truckWeight = wait.poll()
                    possible += truckWeight
                    bridge.offer(truckWeight)
                    cross[index] = true
                    index++
                }
            }



            for (i in cross.indices){
                if (cross[i]){ //다리위에 있는 트럭의 인덱스
                    trucks[i]++
                    if (trucks[i] >= bridge_length && cross[i]){ //다리를 지난경우
                        cross[i] = false
                        possible -= bridge.poll() //다리가 버틸 수 있는 무게 증가
                    }
                }


            }
        }

        return time
    }
}

